G3tHK DB

sanahabhimani opened the pull request:
Currently, this code is a template that I need to heavily iterate upon. It creates a database of two tables: **hkfeeds, hkfields**. 

**hkfeeds** stores information for a .g3 housekeeping file such as: filename, filepath, start time of .g3 file in ctime. 

**hkfields** stores more detailed info for each housekeeping file: field name in each file in the hkfeeds table, and the start and end time for each field

Ideally, I want to split this code into 3 separate scripts, just like the g3tsmurf code. Am currently working on an update HK db script that will periodically (~every hour) update the database (likely after each HK file is complete).

In another iteration, I also want to add more detailed information in the hkfeeds table (such as a description that will tell us which telescope an HK .g3 file is attached to).

kmharrington commented on sotodlib/io/G3tHK.py:
```diff
@@ -0,0 +1,163 @@
+        path to the HK file
+    """
+    __tablename__ = 'hkfeeds'
+    __table_args__ = (
```
make these separate constraints.

kmharrington commented on sotodlib/io/G3tHK.py:
```diff
@@ -0,0 +1,163 @@
+
+    id = db.Column(db.Integer, primary_key=True)
+    fields = relationship("HKFields", back_populates='hkfeed')
+    filename = db.Column(db.String)
```
make `nullable=False, unique=True`

kmharrington commented on sotodlib/io/G3tHK.py:
```diff
@@ -0,0 +1,163 @@
+Session = sessionmaker()
+
+# all database definitions for G3tHK
+class HKFeeds(Base):
```
Feeds = Files?

kmharrington commented on sotodlib/io/G3tHK.py:
```diff
@@ -0,0 +1,163 @@
+
+        return hkfs, starts, ends
+
+    def add_hkfeeds(self, hkarchive_path, hkfile):
```
look at `index_files` and `index_metadata` in `load_smurf` for examples of going through the whole directory. And this is where `min_ctime` and `max_ctime` comes in.

kmharrington commented:
Comments from discussion:
* How many aggregators do we have? Will it still be one per `data_prefix` like we have with the lab data or are we going to be looking somewhere else entirely for the HK data. 
* How will this work for HK fields that we want to be able to index but are site-wide (ex: weather?). Idea: Make one per aggregator and make the loading system able to deal will loading more than one HK instances?

Things the data packaging folks are interested in the short term are for things more closely linked to receivers, az, el, and HWP. But we probably want a sketch of the whole system.

sanahabhimani commented:
Updates as of May 13, 2023:

1. added a new hkagents table—adds an agent for each file, attached with a file_id, and outputs a start and stop time (this is within the length of one .g3 file) of the agent by gathering all starts and stops for each field in that agent, and then only adding the min start time and the max stop time. 
2. updated hkfiles tables to have an aggregator column
3. updated hkfields table to have an agent_id column 

These tables are definitely not perfect:
1. the start and stop for the HKAgents don't yet have a check whether the agent is collecting any useful data. The plan there is to use ocs registry info + hk feed data to essentially check that there's no data for the agent even if it's on
2. I don't think I currently have a way to tell `add_hkfiles` to populate when there's a new .g3 file being written

Am working on a script to query hkdata from these databases with the intention to output that data using hk axismanagers from `hk_utils`. Is that how you sort of envisioned this?

to run the code (on simons1, using old latr data):
```import g3thk
hkarchive_path = '/home/tanay/data/hk/16644'
gethk = g3thk.G3thk(hkarchive_path, db_path = 'sab_trial.db')
gethk.add_hkfiles()
gethk.add_hkagents()
gethk.add_hkfields()```

Table outputs (small snippets... 17 .g3 files with 3956 fields):
HKFiles:
``` select * from hkfiles;
id  filename       global_start_time  path                       aggregator 
--  -------------  -----------------  -------------------------  -----------
1   1664420642.g3  1664420642         /home/tanay/data/hk/16644  observatory
2   1664417033.g3  1664417033         /home/tanay/data/hk/16644  observatory
3   1664406225.g3  1664406225         /home/tanay/data/hk/16644  observatory
4   1664442293.g3  1664442293         /home/tanay/data/hk/16644  observatory
5   1664449514.g3  1664449514         /home/tanay/data/hk/16644  observatory
6   1664409828.g3  1664409828         /home/tanay/data/hk/16644  observatory
```

HKAgents:
```select * from hkagents;
id   file_id  instance_id  alias  start             stop            
---  -------  -----------  -----  ----------------  ----------------
1    1        LSA231C             1664420595.80176  1664424203.18334
2    1        bluefors            1664420278        1664423873      
3    1        LSA22HA             1664420625.1157   1664424225.91071
4    1        LSA22VR             1664420298.00959  1664423897.01007
5    1        LSA22YY             1664420619.89024  1664424195.21633
6    1        dlm                 1664420570.61238  1664424172.84887
```

HKFields:
```select * from hkfields;
id    file_id  agent_id  field                                                   start             end             
----  -------  --------  ------------------------------------------------------ 
1     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch1       1664420599.1231   1664424206.40348
2     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch2       1664420599.1231   1664424206.40348
3     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch3       1664420599.1231   1664424206.40348
4     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch4       1664420599.1231   1664424206.40348
5     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch5       1664420599.1231   1664424206.40348
6     1        11        observatory.pfeiffer.feeds.pressures.pressure_ch6       1664420599.1231   1664424206.40348
7     1        3         observatory.LSA22HA.feeds.temperatures.Channel_12_R     1664420671.94533  1664424214.70841
8     1        3         observatory.LSA22HA.feeds.temperatures.Channel_12_T     1664420671.94533  1664424214.70```

kmharrington reviewed (approved):
we have documentation and usage instructions to write in the future, but we need a version of this now and this has been tested on sat1 data and E2E testing data.
